strawberry/field.py

```diff
@@ -23,129 +23,289 @@
+                    self.python_name, self.origin.__name__
+                )
 
-        self.deprecation_reason = deprecation_reason
+            assert_message = (
+                "Field must have a name by the time the schema is generated"
+            )
+            assert self.python_name is not None, assert_message
 
-    def __call__(self, resolver: _RESOLVER_TYPE) -> "StrawberryField":
-        """Add a resolver to the field"""
+    @classmethod
+    def create(
+        cls,
+        python_name: Optional[str] = None,
+        graphql_name: Optional[str] = None,
+        type_annotation: Optional[Union[StrawberryAnnotation, type]] = None,
+        origin: Optional[Union[Type, Callable, staticmethod, classmethod]] = None,
+        base_resolver: Optional[StrawberryResolver] = None,
+        is_subscription: bool = False,
+        description: Optional[str] = None,
+        deprecation_reason: Optional[str] = None,
+        permission_classes=None,  # type: ignore
+        directives: Sequence[object] = (),
+        dataclass_options: Optional[DataclassFieldSettings] = DEFAULT_FIELD_SETTINGS,
+        is_basic_field: bool = False,
+        # legacy compatibility, will be passed to dataclass_options:
+        default: object = UNSET,
+        default_factory: Union[Callable[[], Any], object] = UNSET,
+    ) -> "StrawberryField":
+        """
+        This is a base method,
+        it will be used by either from_resolver or from_basic_field.
+        users can override this, to provide custom logic for creating the type.
+        it will be called by `strawberry.type`
+        """
+        if permission_classes is None:
+            permission_classes = []
+
+        if default or default_factory:
+            dataclass_options = DataclassFieldSettings(
+                default=default,
+                default_factory=default_factory,
+                is_basic_field=is_basic_field,
+            )
+        type_annotation
+        return cls(
+            python_name=python_name,
+            graphql_name=graphql_name or python_name,
+            type_annotation=type_annotation,
+            origin=origin,
+            is_subscription=is_subscription,
+            description=description,
+            deprecation_reason=deprecation_reason,
+            base_resolver=base_resolver,
+            permission_classes=permission_classes,
+            directives=directives,
+            dataclass_options=dataclass_options,
+            is_basic_field=is_basic_field,
+        )
 
-        # Allow for StrawberryResolvers or bare functions to be provided
-        if not isinstance(resolver, StrawberryResolver):
-            resolver = StrawberryResolver(resolver)
+    @classmethod
+    def evolve_with_resolver(
+        cls,
+        instance: "StrawberryField",
+        base_resolver: StrawberryResolver,
+        origin: Union[Type, Callable, staticmethod, classmethod],
+    ) -> "StrawberryField":
```

Comment:
I'm unclear why we need this and can't use standard inheritance 🤔

Reply:
This is not for overriding `StrawberryField` this is just me refactoring... instead of having the `StrawberryField` built up in so many places we build it only using several functions.

Reply:
mmh, what do you mean exactly?

Reply:
for instance field.origin is determined in `strawberry.type` IIRC it is quite cumbersome IMO.

Reply:
Here:
https://github.com/strawberry-graphql/strawberry/blob/1fc34fa708ebb0d21c6e1a34f86c6202fcfb4783/strawberry/types/type_resolver.py#L116

not only that it is weird that `_get_field` (that is a getter) is messing up with state, it is also unclear to me why wouldn't this logic would live inside `StrawberryField`.

Reply:
That sounds good, but I'm not sure why we need to use these methods and not use the constructor or have a single evaluate call for this 😊

Reply:
> use the constructor
It is impossible with frozen dataclasses.
> have a single evaluate call for this.
there actually is only one evaluate call for this in `StrawberryLazyField` :upside_down_face:  , Wym?

Reply:
do we need to make it frozen?

when is `evolve_with_resolver` called?

Reply:
> do we need to make it frozen?  

I like it this way, keeps things more robust.

> when is evolve_with_resolver called? 

currently [here](https://github.com/strawberry-graphql/strawberry/pull/2117/files#diff-00af9d43c9b59404ba170e72e470939f2a6dd50e2eae24d0e24ef105431e5414R454).Though  I am still trying to figure out where is the best place to call `StrawberryLazyField.evaluate`